### PR TITLE
Add docker executor to help preserve job

### PR DESCRIPTION
**Pull request by mzmssg**

### CODE REVIEW

File: hadoop-ai/hadoop-build/build.sh [hunk 1]
```diff
@@ -31,9 +31,11 @@ git checkout branch-2.9.0
 
 cp /hadoop-2.9.0.gpu-port.patch /hadoop
 cp /HDFS-13773.patch /hadoop
+cp /docker-executor.patch /hadoop
```
Comment (qinchen123) [inline hadoop-ai/hadoop-build/build.sh]: Where this patch comes from ?

Comment (qinchen123) [inline hadoop-ai/hadoop-build/build.sh]: >/docker-executor.patch [](start = 3, length = 22)

This file is under hadoop/hadoop-ai/hadoop-build.   this is node current folder ?

Comment (qinchen123) [reply hadoop-ai/hadoop-build/build.sh]: this file is copy to root folder in dockerfile, Need add some comment here.

---
In reply to: [214570065](https://github.com/Microsoft/pai/pull/1247#discussion_r214570065) [](ancestors = 214570065)

Comment (qinchen123) [inline hadoop-ai/hadoop-build/build.sh]: The file change the Hadoop code which depends on the patch hadoop-2.9.0.gpu-port.20180725.patch, If the dependance patch change, this patch will apply failed.
why not merge this change to the hadoop-2.9.0 . patch ?

Comment (mzmssg) [reply hadoop-ai/hadoop-build/build.sh]: This patch doesn't depend on GPU patch, its change is based on official hadoop,  the conflict possibility is very low.  So I think it is more appropriate of working as a single patch.

File: hadoop-ai/hadoop-build/docker-executor.patch [hunk 2]
```diff
@@ -0,0 +1,114 @@
+diff --git a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-api/src/main/java/org/apache/hadoop/yarn/conf/YarnConfiguration.java b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-api/src/main/java/org/apache/hadoop/yarn/conf/YarnConfiguration.java
+index 96f6c57..1b89e90 100644
+--- a/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-api/src/main/java/org/apache/hadoop/yarn/conf/YarnConfiguration.java
++++ b/hadoop-yarn-project/hadoop-yarn/hadoop-yarn-api/src/main/java/org/apache/hadoop/yarn/conf/YarnConfiguration.java
```
Comment (qinchen123) [inline hadoop-ai/hadoop-build/docker-executor.patch]: >hadoop-yarn/hadoop-yarn-api/src/main/java/org/apache/hadoop/yarn/conf/YarnConfiguration.java [](start = 26, length = 92)

this file may be changed in hadoop-ai patch. then your path will apply failed.

Comment (mzmssg) [reply hadoop-ai/hadoop-build/docker-executor.patch]: For now, a single patch is convenient to fix bugs. As it is stable, I will merge it to hadoop-ai.

File: pai-management/bootstrap/hadoop-node-manager/hadoop-node-manager-configuration/nodemanager-generate-script.sh [hunk 3]
```diff
@@ -65,3 +68,7 @@ sed  -i "s/{cpu_vcores}/${cpu_vcores}/g" $HADOOP_CONF_DIR/yarn-site.xml
 
 # Backup hadoop configuration for job (Spark) use
 cp $HADOOP_CONF_DIR/* /hadoop-configuration-for-jobs/
+
+# Copy docker .config file for yarn-container
+cp -a /root/.docker /docker-config/
```
Comment (ydye) [inline pai-management/bootstrap/hadoop-node-manager/hadoop-node-manager-configuration/nodemanager-generate-script.sh]: What purpose to add this?

File: pai-management/bootstrap/hadoop-node-manager/hadoop-node-manager-configuration/yarn-site.xml [hunk 4]
```diff
@@ -272,5 +272,44 @@
...
+  <name>yarn.nodemanager.container-executor.class</name>
+  <value>org.apache.hadoop.yarn.server.nodemanager.DockerContainerExecutor</value>
+  <description>
+     This is the container executor setting that ensures that all
+jobs are started with the DockerContainerExecutor.
+  </description>
+</property>
+
+<property>
+  <name>yarn.nodemanager.docker-container-executor.image-name</name>
```
Comment (ydye) [inline pai-management/bootstrap/hadoop-node-manager/hadoop-node-manager-configuration/yarn-site.xml]: You need hadoop-run image to run the am container?

File: pai-management/bootstrap/hadoop-node-manager/hadoop-node-manager-configuration/nodemanager-generate-script.sh (hunk 3 above)
Comment (mzmssg) [reply pai-management/bootstrap/hadoop-node-manager/hadoop-node-manager-configuration/nodemanager-generate-script.sh]: I thought it is for private registry support.
But after checking the code, the private job container is authenticated in yarnContainerScript, this volume is redundant and should be removed.

Another question, what's the use of mount it to node manager?

Comment (ydye) [reply pai-management/bootstrap/hadoop-node-manager/hadoop-node-manager-configuration/nodemanager-generate-script.sh]: Can we remove this line and corresponding volume？

Comment (mzmssg) [reply pai-management/bootstrap/hadoop-node-manager/hadoop-node-manager-configuration/nodemanager-generate-script.sh]: I'm not sure whether other features depend on this configmap, in my test, removing this volume from nm yaml seems no difference

**Pull request merged**